File: pythonX/hard/h004_minimum_window_substring.py
"""
H004: Minimum Window Substring

Summary:
Find minimum window in s containing all characters from t.

Requirements:
Return shortest substring of s that contains all chars from t.

Approach:
Use sliding window with two dicts for character frequencies.

Data Structures:
- dict[str, int]: frequency maps for t and window
- Two pointers for sliding window

Big O Analysis:
Time Complexity: O(m + n) where m=len(s), n=len(t)
Space Complexity: O(m + n)
"""

from collections import Counter


def min_window(s: str, t: str) -> str:
    """
    TODO: Find minimum window substring
    1. Create frequency map for t
    2. Initialize window map, left=0, min_len, min_start
    3. Track required chars and have chars
    4. Expand with right pointer:
       - Add char to window
       - If frequency matches, increment have
    5. When have == required, shrink with left:
       - Update minimum if smaller
       - Remove chars from left
    6. Return substring or ""
    """
    pass


if __name__ == "__main__":
    # Test Cases
    
    print("Test 1: Basic case")
    print(f"s='ADOBECODEBANC', t='ABC': '{min_window('ADOBECODEBANC', 'ABC')}'")  # Expected: BANC
    
    print("\nTest 2: Exact match")
    print(f"s='a', t='a': '{min_window('a', 'a')}'")  # Expected: a
    
    print("\nTest 3: No solution")
    print(f"s='a', t='aa': '{min_window('a', 'aa')}'")  # Expected: ""
    
    print("\nTest 4: Single char window")
    print(f"s='ab', t='b': '{min_window('ab', 'b')}'")  # Expected: b
    
    print("\nTest 5: Multiple valid windows")
    print(f"s='cabwefgewcwaefgcf', t='cae': '{min_window('cabwefgewcwaefgcf', 'cae')}'")  # Expected: cwae or similar

File: src/com/example/javas/hard/H004_MinimumWindowSubstring.java
/*
HARD PROBLEM (Simplified for 25 min):
Minimum Window Substring

REQUIREMENTS:
- Given string s and string t, find the minimum window in s that contains all characters from t
- Return the minimum window substring, or "" if no such window exists
- If multiple windows exist, return any one

APPROACH:
- Use sliding window with two HashMaps
- Expand window to include all characters from t
- Shrink window to find minimum length

DATA STRUCTURES:
- HashMap<Character, Integer>: frequency maps for t and current window

TIME COMPLEXITY: O(m + n) where m = s.length, n = t.length
SPACE COMPLEXITY: O(m + n)
*/

package com.example.javas.hard;

import java.util.*;

public class H004_MinimumWindowSubstring {
    
    public static String minWindow(String s, String t) {
        // TODO: Implement minimum window finder
        // 1. Create frequency map for t
        // 2. Initialize window map, left = 0, minLen = Integer.MAX_VALUE, minStart = 0
        // 3. Track how many unique chars from t are satisfied (have)
        // 4. Expand with right pointer:
        //    - Add char to window map
        //    - If char frequency matches t, increment have
        // 5. When have == need, shrink with left pointer:
        //    - Update minimum window if smaller
        //    - Remove char from window, decrement have if needed
        // 6. Return substring or ""
        
        return "";
    }
    
    // TEST CASES
    public static void main(String[] args) {
        System.out.println("s=\"ADOBECODEBANC\", t=\"ABC\" -> \"" + minWindow("ADOBECODEBANC", "ABC") + "\"");
        System.out.println("s=\"a\", t=\"a\" -> \"" + minWindow("a", "a") + "\"");
        System.out.println("s=\"a\", t=\"aa\" -> \"" + minWindow("a", "aa") + "\"");
        System.out.println("s=\"ab\", t=\"b\" -> \"" + minWindow("ab", "b") + "\"");
    }
}
